### DISCUSSION TITLE
Corruption Verification

### DISCUSSION THREAD

**Discussion by KahinaD** [samuel-lucas6/Kryptor#53] [Q&A]
### Description
This not so much a bug as it is me trying to figure out what went wrong in the decryption. I was decrypting my files so I could update from v3.0.4 to 4.0.1 when I found I wasn't able to decrypt about half the files. They just throw this error: 

Deriving encryption key from password...
Decrypting filename.kryptor => filename...
filename.kryptor - Error: CryptographicException - Unable to decrypt the file.

I am using the same keyfile as it worked on the other files but there were a few that it wouldn't work with. I figure it's just bitrot or something. I had the files on a USB that was only accessed weekly running NTFS. Is there anything I can do to recover the files or at least make them decryptable (hex editor and screwing with the text maybe, idk)? I can remake the files if need be so there's no data loss, just wondering if there's anything I can do in order to not have to remake them.

I am on Windows 10, 21H2 (though I doubt this to be an OS issue, can try in Linux if necessary).

**Comment by samuel-lucas6** [comment]
That's not good; I'm sorry to hear that. However, as v3.0.4 is quite old, I've moved this to a discussion.

`CryptographicException` implies either the key is wrong or the files have been modified. There's no way for me to tell from the error alone. If you tried decryption in a debugger, you might be able to see whether the exception throws immediately or midway through a file. However, that requires using and being familiar with something like Visual Studio or JetBrains Rider. It also likely won't help you recover your files because it still reveals little information.

Was the keyfile randomly generated? If so, is it still read-only? Could the keyfile have been modified in the time between decrypting the other files and decrypting these files, or did you do all of the decryption at the same time? Is it possible some other method of encryption or a different key was used? Are the files that failed to decrypt still read-only? Are you using v3.0.4 for decryption?

I'd recommend updating to [v3.1.1](https://github.com/samuel-lucas6/Kryptor/releases/tag/v3.1.1) for further attempts to address a decryption bug. The only other thing you could do is look at the [human readable file headers](https://web.archive.org/web/20210921211237/https://www.kryptor.co.uk/technical-details#header-structure) (first two) in a hex editor.

**Comment by KahinaD** [comment]
The keyfile used was an image file that couldn't have been modified since I decrypted a bunch of files all at the same time and it was only a few that threw the error. I made a simple script to bulk decrypt them and it did work for most (tried it manually later with the broken ones, same issue). 

Just tested them on 3.1.1 and it still has the same issue. I think I'll try messing with the program on VS later and see if I can find out exactly what the exception is. I still figuring it has to be an issue with the encrypted files. Like a few bytes or something changed on it at some point breaking it. Still not a big loss because I've since managed to acquire backups of nearly every file. Thanks for the quick response.

**Reply by samuel-lucas6** [reply]
Thanks for the info. That's good to hear.
